### PR TITLE
[SPARK-23217][ML] Add cosine distance measure to ClusteringEvaluator

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR provided an implementation of ClusteringEvaluator using the cosine distance measure.
This allows to evaluate clustering results created using the cosine distance, introduced in SPARK-22119.

In the corresponding JIRA, there is a design document for the algorithm implemented here.

## How was this patch tested?

Added UT which compares the result to the one provided by python sklearn.

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @yanboliang

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 1]
```diff
@@ -111,6 +129,53 @@ object ClusteringEvaluator
...
+      clusterIds: Set[Double],
+      pointClusterId: Double,
+      pointClusterNumOfPoints: Long,
+      averageDistanceToCluster: (Double) => Double): Double = {
+    // Here we compute the average dissimilarity of the current point to any cluster of which the
+    // point is not a member.
+    // The cluster with the lowest average dissimilarity - i.e. the nearest cluster to the current
+    // point - s said to be the "neighboring cluster".
+    var neighboringClusterDissimilarity = Double.MaxValue
+    clusterIds.foreach {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: What about `clusterIds.filter(c != pointClusterId).map(averageDistanceToCluster).min` (except that it needs to deal with the case that `filter` returns no elements

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 2]
```diff
@@ -111,6 +129,53 @@ object ClusteringEvaluator
 }
 
 
+private[evaluation] class Silhouette {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: should this be abstract?

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 3]
```diff
@@ -421,13 +460,220 @@ private[evaluation] object SquaredEuclideanSilhouette {
...
+   * @return A [[scala.collection.immutable.Map]] which associates each cluster id to a
+   *         [[ClusterStats]] object (which contains the precomputed values `N` and
+   *         `$\Omega_{\Gamma}$`).
+   */
+  def computeClusterStats(df: DataFrame, predictionCol: String): Map[Double, ClusterStats] = {
+    val numFeatures = df.select(col(normalizedFeaturesColName)).first().getAs[Vector](0).size
+    val clustersStatsRDD = df.select(
+      col(predictionCol).cast(DoubleType), col(normalizedFeaturesColName))
+      .rdd
+      .map { row => (row.getDouble(0), row.getAs[Vector](1)) }
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: Does this work as `.as[(Double,Vector)]` before `.rdd`?

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 4]
```diff
@@ -421,13 +460,220 @@ private[evaluation] object SquaredEuclideanSilhouette {
...
+      combOp = {
+        case ((normalizedFeaturesSum1, numOfPoints1), (normalizedFeaturesSum2, numOfPoints2)) =>
+          BLAS.axpy(1.0, normalizedFeaturesSum2, normalizedFeaturesSum1)
+          (normalizedFeaturesSum1, numOfPoints1 + numOfPoints2)
+      }
+    )
+
+    clustersStatsRDD
+      .collectAsMap()
+      .mapValues {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: What about `mapValues` on the RDD before collecting or does that not work?

File: mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala [hunk 5]
```diff
@@ -66,16 +66,38 @@ class ClusteringEvaluatorSuite
     assert(evaluator.evaluate(irisDataset) ~== 0.6564679231 relTol 1e-5)
   }
 
-  test("number of clusters must be greater than one") {
-    val singleClusterDataset = irisDataset.where($"label" === 0.0)
+  /*
+    Use the following python code to load the data and evaluate it using scikit-learn package.
```
Comment (srowen) [inline mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala]: Nit: I think you can use the standard block comment syntax here with leading asterisks

**Review by srowen — CHANGES_REQUESTED** [review]
A few minor code questions to start

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 1 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: we don't need to deal with that case, because it is asserted that `clusterIds.size > 1`

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 3 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: yes, I will switch to it

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 4 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: that would work too, but it is quite pointless up to me, since here we are dealing with the clusters and they are assumed not to be many... usually they are << 100....

File: mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala (hunk 5 above)
Comment (mgaido91) [reply mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala]: this is the same as for `squaredEuclidean` where this format was used according to  https://github.com/apache/spark/pull/18538#discussion_r131100309

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 6]
```diff
@@ -84,18 +81,39 @@ class ClusteringEvaluator @Since("2.3.0") (@Since("2.3.0") override val uid: Str
...
 
-  setDefault(metricName -> "silhouette")
+  /**
+   * param for distance measure to be used in evaluation
+   * (supports `"squaredEuclidean"` (default), `"cosine"`)
+   * @group param
+   */
+  @Since("2.4.0")
+  val distanceMeasure: Param[String] = {
+    val allowedParams = ParamValidators.inArray(Array("squaredEuclidean", "cosine"))
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: You don't need to change this, but it occurs to me that on lots of the parameters that take discrete values, the error message could reference the same array of values the validator uses, to make sure they're always consistent.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 3 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: I'm neutral on it. I don't think importing implicits is a big deal either way.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 7]
```diff
@@ -421,13 +453,220 @@ private[evaluation] object SquaredEuclideanSilhouette {
...
+  private[this] val normalizedFeaturesColName = "normalizedFeatures"
+
+  /**
+   * This method registers the class
+   * [[org.apache.spark.ml.evaluation.CosineSilhouette.ClusterStats]]
+   * for kryo serialization.
+   *
+   * @param sc `SparkContext` to be used
+   */
+  def registerKryoClasses(sc: SparkContext): Unit = {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: This duplicates a method in ClusteringEvaluator right? I wonder if this can happen just once. It's OK if it registers a bunch of classes, not all of which will be used.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 4 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: OK sure, it's not worth another distributed stage.

File: mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala (hunk 5 above)
Comment (srowen) [reply mllib/src/test/scala/org/apache/spark/ml/evaluation/ClusteringEvaluatorSuite.scala]: I see, the idea is to make it more copy-pasteable. That's fine.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 8]
```diff
@@ -111,6 +129,46 @@ object ClusteringEvaluator
...
+
+    // adjustment for excluding the node itself from the computation of the average dissimilarity
+    val currentClusterDissimilarity = if (pointClusterNumOfPoints == 1) {
+      0
+    } else {
+      averageDistanceToCluster(pointClusterId) * pointClusterNumOfPoints /
+        (pointClusterNumOfPoints - 1)
+    }
+
+    (currentClusterDissimilarity compare neighboringClusterDissimilarity).signum match {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: Is this just expressing ...

```
if (currentClusterDissimilarity < neighboringClusterDissimilarity) {
  ...
} else if (currentClusterDissimilarity > neighboringClusterDissimilarity) {

} else {
  ...
}
```

That seems more straightforward if that's all it is, to my eyes. This has postfix notation, signum, match statement

**Review by srowen — COMMENTED** [review]
More minor comments for fun but this is looking OK to me already.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 7 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: sorry, I am not sure I can get what you mean. Which method is this duplicating? The registration happens once since there is a flag for it...

**Comment by mgaido91** [comment]
test failures are unrelated and caused by SPARK-23234

**Comment by mgaido91** [comment]
retest this please

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 7 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: There are two methods like this in each implementation of the Silhouette class. I was wondering if there can be just one place where all relevant classes are registered. It may not be possible without, say, making the superclass refer to its subclasses, and that may not be worth it just to make a single registerKryoClasses method.

**Review by srowen — APPROVED** [review]
That's looking good to me. I wonder if @jkbradley @MLnick @viirya have input?

**Comment by mgaido91** [comment]
kindly ping @jkbradley @MLnick @viirya

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 9]
```diff
@@ -421,13 +456,220 @@ private[evaluation] object SquaredEuclideanSilhouette {
...
+      sc.getConf.registerKryoClasses(
+        Array(
+          classOf[CosineSilhouette.ClusterStats]
+        )
+      )
+      kryoRegistrationPerformed = true
+    }
+  }
+
+  case class ClusterStats(normalizedFeatureSum: Vector, numOfPoints: Long)
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: Back on this -- how about just using a Tuple2 of Vector, Long? no new class to register.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 10]
```diff
@@ -421,13 +456,220 @@ private[evaluation] object SquaredEuclideanSilhouette {
...
+
+    bClustersStatsMap.destroy()
+
+    silhouetteScore
+  }
+}
+
+
+/**
+ * The algorithm which is implemented in this object, instead, is an efficient and parallel
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: From skimming this I wasnt clear why you can only implement it this way for cosine similarity?

**Review by srowen — COMMENTED** [review]
This doesn't have any of the same question about normalizing the input for cosine similarity, right?

This is looking fine. Absent any other comments I think it can be merged.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 10 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: there are 2 reason of my sentence, let me know if it is not clear:
 1. the whole algorithm (all the math steps described) assumes that we are using the cosine distance; for a different distance measure, the algorithm is not valid, even though you can use the same approach to do something similar (as it is done for the squared Euclidean, above);
 2. for some distance measure it is not possible to use this approach - ie. aggregate the score of the points - because math doesn't allow that. For instance, you can't use this approach with the Euclidean distance, since the sqrt prevents any possible aggregation.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 9 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: I just thought it was clearer. Do you think it is better to use a Tuple2 for this and Tuple3 for `SquaredEuclideanSilhouette`?

Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: I think it avoids another class to deal with registering, which is the real cost. A one-off 2-field class is about as clear in the case of a pair. For a Tuple3, not as clear. Up to your taste

**Comment by srowen** [comment]
Merged to master

**Comment by mgaido91** [comment]
@srowen sorry, this needs also to add the parameter to the python API; given our discussion on JIRA, what should I do? Create a follow-up PR or a new ticket on JIRA?
Thanks.

**Comment by srowen** [comment]
I think you can just add a follow-up
